### PR TITLE
Add cache for get_current_hosts_in_room

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/state.py [hunk 1]
```diff
@@ -176,6 +176,17 @@ def get_current_user_in_room(self, room_id, latest_event_ids=None):
         defer.returnValue(joined_users)
 
     @defer.inlineCallbacks
+    def get_current_hosts_in_room(self, room_id, latest_event_ids=None):
+        if not latest_event_ids:
+            latest_event_ids = yield self.store.get_latest_event_ids_in_room(room_id)
+        logger.debug("calling resolve_state_groups from get_current_user_in_room")
```
Comment (NegativeMjark) [inline synapse/state.py]: Is this debug logging still correct?

File: synapse/storage/roommember.py [hunk 2]
```diff
@@ -482,6 +482,44 @@ def _is_host_joined(self, room_id, host, state_group, current_state_ids):
...
+            # state group, i.e. we need to make sure that calls with a state_group
+            # of None don't hit previous cached calls with a None state_group.
+            # To do this we set the state_group to a new object as object() != object()
+            state_group = object()
+
+        return self._get_joined_hosts(
+            room_id, state_group, state_ids
+        )
+
+    @cachedInlineCallbacks(num_args=3)
```
Comment (NegativeMjark) [inline synapse/storage/roommember.py]: Is this number of arguments here correct?

File: synapse/storage/roommember.py [hunk 3]
```diff
@@ -482,6 +482,44 @@ def _is_host_joined(self, room_id, host, state_group, current_state_ids):
...
+                except:
+                    logger.warn("state_key not user_id: %s", state_key)
+                    continue
+
+                if host in joined_hosts:
+                    continue
+
+                event = yield self.get_event(event_id, allow_none=True)
+                if event and event.content["membership"] == Membership.JOIN:
+                    joined_hosts.add(host)
```
Comment (NegativeMjark) [inline synapse/storage/roommember.py]: Is it worth interning the host strings here?

**Comment by NegativeMjark** [comment]
Are there more places we can use this function?

If I recall correctly lots of places in typing/presence/etc.. list the servers that are in a room.

**Comment by erikjohnston** [comment]
> Are there more places we can use this function?
> If I recall correctly lots of places in typing/presence/etc.. list the servers that are in a room.

I think those places try and get the remote hosts *and* local users

**Pull request merged**